### PR TITLE
Add outline of constraint framework

### PR DESCRIPTION
**Pull request by maxsmythe**
This is a rough outline of what I'm thinking for the framework.

Comments most welcome.

Most people are probably looking for the ConstraintTemplate outline:

constraint/pkg/apis/templates/v1alpha1/constrainttemplate_types.go

Signed-off-by: Max Smythe <smythe@google.com>

### CODE REVIEW

File: constraint/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# Build the manager binary
```
Comment (tsandall) [inline constraint/Dockerfile]: Nit: Is this file needed?

File: constraint/Gopkg.toml [hunk 2]
```diff
@@ -0,0 +1,37 @@
+required = [
```
Comment (tsandall) [inline constraint/Gopkg.toml]: Nit: Can we reduce this to the minimum required dependencies? It looks like this is copied from Gatekeeper. Would be nice to keep the library as lightweight as possible.

File: constraint/Makefile [hunk 3]
```diff
@@ -0,0 +1,55 @@
+
```
Comment (tsandall) [inline constraint/Makefile]: Nit: Same comments as on the go deps and Dockerfile apply here.

File: constraint/pkg/apis/templates/v1alpha1/constrainttemplate_types_test.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	"testing"
+
+	"github.com/onsi/gomega"
+	"golang.org/x/net/context"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func TestStorageConstraintTemplate(t *testing.T) {
+	key := types.NamespacedName{
```
Comment (tsandall) [inline constraint/pkg/apis/templates/v1alpha1/constrainttemplate_types_test.go]: Is this code generated or is there some desire to use gomega? I'd prefer to just use the Go standard library for testing and not rely on one of these frameworks.

File: constraint/Dockerfile (hunk 1 above)
Comment (maxsmythe) [reply constraint/Dockerfile]: Probably not. Let me go through and purge things a little more carefully

File: constraint/Gopkg.toml (hunk 2 above)
Comment (maxsmythe) [reply constraint/Gopkg.toml]: Required files are from kubebuilder. I can look and see what we may be able to drop as we are only using bits of kubebuilder here.

File: constraint/pkg/apis/templates/v1alpha1/constrainttemplate_types_test.go (hunk 4 above)
Comment (maxsmythe) [reply constraint/pkg/apis/templates/v1alpha1/constrainttemplate_types_test.go]: This is generated code.

**Comment by maxsmythe** [comment]
Should have pruned extra libraries with this last amend

File: constraint/Gopkg.toml (hunk 2 above)
Comment (maxsmythe) [reply constraint/Gopkg.toml]: removed some extraneous libs

File: constraint/Makefile (hunk 3 above)
Comment (maxsmythe) [reply constraint/Makefile]: Cleaned

File: constraint/pkg/apis/templates/v1alpha1/constrainttemplate_types_test.go (hunk 4 above)
Comment (tsandall) [reply constraint/pkg/apis/templates/v1alpha1/constrainttemplate_types_test.go]: OK. If let's merge for now and then revisit. If we don't need to take a dependency on gomega, that would be preferable.
